modules/apps/27-interchain-accounts/types/codec_test.go

```diff
@@ -38,118 +41,388 @@ func (mockSdkMsg) GetSigners() []sdk.AccAddress {
+				}
+
+				for i, msg := range msgs {
+					suite.Require().Equal(msgs[i], msg)
+				}
+			})
+		}
+
+		// test serializing non sdk.Msg type
+		bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []proto.Message{&banktypes.MsgSendResponse{}}, encoding)
+		suite.Require().NoError(err)
+		suite.Require().NotEmpty(bz)
+
+		// test deserializing unknown bytes
+		_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, bz, encoding)
+		suite.Require().Error(err) // unregistered type
+
+		// test deserializing unknown bytes
+		msgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []byte("invalid"), types.EncodingProtobuf)
+		suite.Require().Error(err)
+		suite.Require().Empty(msgs)
+	}
 }
 
 // unregistered bytes causes amino to panic.
 // test that DeserializeCosmosTx gracefully returns an error on
 // unsupported amino codec.
-func (suite *TypesTestSuite) TestDeserializeAndSerializeCosmosTxWithAmino() {
+func (suite *TypesTestSuite) TestProtoDeserializeAndSerializeCosmosTxWithAmino() {
 	cdc := codec.NewLegacyAmino()
 	marshaler := codec.NewAminoCodec(cdc)
 
-	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}})
+	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(msgs)
 
-	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0})
+	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(bz)
 }
+
+func (suite *TypesTestSuite) TestCosmwasmDeserializeCosmosTx() {
+	var cwBytes []byte
+	var protoMessages []proto.Message
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success: single msg from cosmwasm",
+			func() {
+				cwBytes = []byte{123, 34, 109, 101, 115, 115, 97, 103, 101, 115, 34, 58, 91, 123, 34, 64, 116, 121, 112, 101, 34, 58, 34, 47, 99, 111, 115, 109, 111, 115, 46, 98, 97, 110, 107, 46, 118, 49, 98, 101, 116, 97, 49, 46, 77, 115, 103, 83, 101, 110, 100, 34, 44, 34, 102, 114, 111, 109, 95, 97, 100, 100, 114, 101, 115, 115, 34, 58, 34, 99, 111, 115, 109, 111, 115, 49, 55, 100, 116, 108, 48, 109, 106, 116, 51, 116, 55, 55, 107, 112, 117, 104, 103, 50, 101, 100, 113, 122, 106, 112, 115, 122, 117, 108, 119, 104, 103, 122, 117, 106, 57, 108, 106, 115, 34, 44, 34, 116, 111, 95, 97, 100, 100, 114, 101, 115, 115, 34, 58, 34, 99, 111, 115, 109, 111, 115, 49, 55, 100, 116, 108, 48, 109, 106, 116, 51, 116, 55, 55, 107, 112, 117, 104, 103, 50, 101, 100, 113, 122, 106, 112, 115, 122, 117, 108, 119, 104, 103, 122, 117, 106, 57, 108, 106, 115, 34, 44, 34, 97, 109, 111, 117, 110, 116, 34, 58, 91, 123, 34, 100, 101, 110, 111, 109, 34, 58, 34, 98, 97, 110, 97, 110, 97, 115, 34, 44, 34, 97, 109, 111, 117, 110, 116, 34, 58, 34, 49, 48, 48, 34, 125, 93, 125, 93, 125}
```

Comment:
Again if this can't be generated in the test function, document what the message is and how you generated it for each byte array

Reply:
It could be generated here but I wanted it to be authentic. If you wrap these in a `string()`, you'll see the actual json string, so I don't think I need to document this, however, I understand if you'd like me to use string here instead so it is more human readable?

Reply:
I made tests human readable